test/functional/test_framework/test_node.py

```diff
@@ -267,6 +284,96 @@ def assert_debug_log(self, expected_msgs):
                 if re.search(re.escape(expected_msg), log, flags=re.MULTILINE) is None:
                     self._raise_assertion_error('Expected message "{}" does not partially match log:\n\n{}\n\n'.format(expected_msg, print_log))
 
+    @contextlib.contextmanager
+    def profile_with_perf(self, profile_name):
+        """
+        Context manager that allows easy profiling of node activity using `perf`.
+
+        Perf will sample the running node and will generate profile data in the node's
+        datadir. The profile data can then be presented using `perf report` or a graphical
+        tool like `hotspot`<https://github.com/KDAB/hotspot>.
+
+        To see useful textual output, run
+
+            perf report -i /path/to/datadir/node-0.perf.data --stdio | c++filt | less
+
+        See also:
+
+            Installing perf: https://askubuntu.com/q/50145
+            Perf examples: http://www.brendangregg.com/perf.html
+
+        Args:
+
+            profile_name (str): This string will be appended to the
+                profile data filename generated by perf.
+
+        Usage:
+
+            with node.profile_with_perf("send-big-msgs"):
+                # Perform activity on the node you're interested in profiling, e.g.:
+                for _ in range(10000):
+                    node.p2p.send_message(some_large_message)
+        """
+        subp = self._start_perf(profile_name)
+
+        yield
+
+        if subp:
+            self._stop_perf(profile_name)
+
+    def _start_perf(self, profile_name):
+        """Start a perf process to profile this node.
+
+        Returns the subprocess running perf.
+        """
+        subp = None
+
+        def test_success(cmd):
+            try:
+                return subprocess.run(
```

Comment:
We don't support this function yet: https://docs.python.org/3.4/library/subprocess.html#subprocess.run